src/foam/dao/QueryCachingDAODecorator.js

```diff
@@ -0,0 +1,156 @@
+              sink.put(self.cache[key][idx]);
+            }
+
+            resolve(sink);
+          });
+        });
+      });
+    },
+
+    // Remove invalidates cache and is forwarded to the source.
+    function remove_(x, o) {
+      this.cache = {};
+      this.clearProperty('daoCount_');
+      return this.delegate.remove_(x, o);
+    },
+
+    // RemoveAll invalidates cache and is forwarded to the source.
+    function removeAll_(x, skip, limit, order, predicate) {
+      this.cache = {};
+      this.clearProperty('daoCount_');
+      this.delegate.removeAll_(x, skip, limit, order, predicate);
+    },
+
+    function cmd_(x, obj) {
+      if ( obj == this.CLEAR_CACHE ) {
+        this.cache = {};
+        this.clearProperty('daoCount_');
+      } else {
+        this.SUPER(x, obj);
+      }
+    },
+
+    function fillCache_(self, key, requestStartIdx, requestEndIdx, x, sink, order, predicate) {
+      // Pre-check cache for elements
+      let startIdx = -1;
+      let endIdx = -1;
+      let hasMissingData = false;
+
+      if (self.cache[key]) {
+        // Cycle through exising cached elements to verify all requested are present
+        for ( let idx = requestStartIdx; idx < requestEndIdx; idx++ ) {
+          if ( ! self.cache[key][idx] ) {
+            if ( ! hasMissingData ) {
+              // Found start of missing data withing requested block
+              hasMissingData = true;
+              startIdx = idx;
+            }
+
+            // Make sure we have the last index of a missing data element within requested block
+            endIdx = idx + 1;
+          }
+        }
+      } else {
+        // No data present load entire block
+        hasMissingData = true;
+        startIdx = requestStartIdx;
+        endIdx = requestEndIdx;
+        self.cache[key] = [];
+      }
+
+      if (hasMissingData) {
+        console.log('******** QUERYCACHE*** HAS MISSING DATA ***: key: ' + key + 'daoCount: ' + self.daoCount_ + ' startIdx: ' + startIdx + ' endIdx: ' + endIdx);
+        return this.delegate.select_(x, sink, startIdx, 1 + endIdx - startIdx, order, predicate).then(function (result) {
+          // Point to the appropriate array source
+//          let array = foam.mlang.sink.Projection.isInstance(sink) ? result.projectionWithClass : result.array;
+
+          // Update cache with missing data
+          for (let idx = 0; idx < result.array.length; idx++) {
+            if ( ! self.cache[key][startIdx + idx]) {
+              self.cache[key][startIdx + idx] = result.array[idx];
+            }
+          }
+        });
+      }
+
+      return Promise.resolve();
+    },
+
+    function refreshDaoCount_(self) {
+      // If we have not retrieved the dao count previously do it now
+      if ( ! self.hasOwnProperty('daoCount_') ) {
+        return self.delegate.select_(foam.mlang.sink.Count.create()).then( function(count) {
+          self.daoCount_ = count.array.length;
```

Comment:
This doesn't make sense, since Count only has one property: 'value', not 'array', which would suggest it was an ArraySink being returned.

Reply:
![Screen Shot 2021-08-16 at 9 29 38 AM](https://user-images.githubusercontent.com/85950587/129571632-b2e6ad5e-b135-4f9a-84b4-db488e6a1006.png)

count.cls_ at this breakpoint 
Object {prototype: Object, axiomMap_: Object, private_: Object, count_: 1479, id: "foam.dao.ArraySink", ...}

Reply:
I was only passing Count sink to the select_ which has x as first parameter ... now count.value works and have fixed the code to use it